gcstorage/gcstorage.go

```diff
@@ -0,0 +1,253 @@
+	r, err := obj.NewReader(ctx)
+	if err != nil && errors.Is(err, storage.ErrObjectNotExist) {
+		return errReturnObj, nil, false, nil
+	} else if err != nil {
+		return errReturnObj, nil, false, err
+	}
+	defer r.Close()
+
+	data, err := io.ReadAll(r)
+	if err != nil {
+		return errReturnObj, nil, false, err
+	}
+
+	value, err := deserialize[cacheObject[V]](data)
+	if err != nil {
+		return errReturnObj, nil, false, err
+	}
+
+	if value.ExpiresAt != (time.Time{}) && value.ExpiresAt.Before(time.Now()) {
+		return errReturnObj, nil, false, nil
+	}
+
+	ttl := value.ExpiresAt.Sub(time.Now())
+	return value.Object, &ttl, true, nil
+}
+
+func (g *GCStorage[V]) BatchSet(ctx context.Context, keys []string, values []V) error {
+	return g.BatchSetEx(ctx, keys, values, g.defaultKeyExpiry)
+}
+
+func (g *GCStorage[V]) BatchSetEx(ctx context.Context, keys []string, values []V, ttl time.Duration) error {
+	for i, key := range keys {
+		if err := g.SetEx(ctx, key, values[i], ttl); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (g *GCStorage[V]) BatchGet(ctx context.Context, keys []string) ([]V, []bool, error) {
+	var out []V
+	var exists []bool
+	for _, key := range keys {
+		value, exist, err := g.Get(ctx, key)
+		if err != nil {
+			return nil, nil, err
+		}
+		out = append(out, value)
+		exists = append(exists, exist)
+	}
+	return out, exists, nil
+}
+
+func (g *GCStorage[V]) Delete(ctx context.Context, key string) error {
+	obj := g.bucketHandle.Object(g.keyPrefix + key)
+	return obj.Delete(ctx)
+}
+
+func (g *GCStorage[V]) DeletePrefix(ctx context.Context, keyPrefix string) error {
+	objIt := g.bucketHandle.Objects(ctx, &storage.Query{
+		Prefix: g.keyPrefix + keyPrefix,
+	})
+
+	for {
+		objAttrs, err := objIt.Next()
+		if err != nil && errors.Is(err, iterator.Done) {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		if err = g.bucketHandle.Object(objAttrs.Name).Delete(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (g *GCStorage[V]) ClearAll(ctx context.Context) error {
+	objIt := g.bucketHandle.Objects(ctx, &storage.Query{
+		Prefix: g.keyPrefix,
+	})
+
+	for {
+		objAttrs, err := objIt.Next()
+		if err != nil && errors.Is(err, iterator.Done) {
+			break
+		} else if err != nil {
+			return err
+		}
+
+		if err = g.bucketHandle.Object(objAttrs.Name).Delete(ctx); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (g *GCStorage[V]) GetOrSetWithLock(ctx context.Context, key string, getter func(context.Context, string) (V, error)) (V, error) {
+	var out V
+	return out, ErrNotSupported
+}
+
+func (g *GCStorage[V]) GetOrSetWithLockEx(ctx context.Context, key string, getter func(context.Context, string) (V, error), ttl time.Duration) (V, error) {
+	var out V
+	return out, ErrNotSupported
+}
+
+func serialize[V any](value V) ([]byte, error) {
+	return json.Marshal(value)
+}
+
+func deserialize[V any](data []byte) (V, error) {
+	var out V
+	err := json.Unmarshal(data, &out)
+	return out, err
```

Comment:
This is just personal preference, i like to explicitly return nil error on end of the functions.
```suggestion
	if err := json.Unmarshal(data, &out); err != nil {
		return *new(V), err
	}

	return out, nil

```